### ISSUE TITLE
Set target folder of downloads

### ISSUE THREAD

**Issue by Lathostiran** [downthemall/downthemall#2]
Successfully pulled from multiple websites. was able to use the mask and *batch* to create directories.

but for new users the default behavior is that it just dumps to downloads, potentially spamming the users file system.

could a "download to..." dropdown/preference be added?

**Comment by nmaier** [comment]
Sadly it cannot at the moment. I'd love that too. But WebExtensions simply do not allow to do that; the downloads API will always dump to the browser configured download folder.

You can however use relative paths (no ".." tho) in the mask, like `dta/*num*/*name*.*ext*` to create subfolders within your download directory.

**Comment by Lathostiran** [comment]
then perhaps changing the default installation mask to "dta/*name*.*ext*" to show the user the capability. 

those who want something else will be editing the field anyway and the starting user gets an organized batch.

**Comment by nmaier** [comment]
Well, not the very default mask, but I can certainly add an item to the defaults of the dropdown.

**Comment by Lathostiran** [comment]
ah, i had missed the dropdown. too excited to test the return of a good utility.

Thanks for your replys.

**Comment by nmaier** [comment]
[Bugzilla] https://bugzilla.mozilla.org/show_bug.cgi?id=1342563

**Comment by kpedro88** [comment]
If I may make a suggestion: it would be especially useful to have the download path as a separate box from the mask (a la the old DTA).

This would, with a little tweak, allow savvy users to entirely circumvent the new restrictions on download folders.

We just need these two lines in an elevated command prompt:
```
cd %USERPROFILE%\Downloads
mklink /D "C" "C:\"
```

Now, we can specify a path "C\..." (relative to the Downloads folder) and access the entire drive. This has to be repeated for other drive letters, but it's fairly straightforward.

**Comment by nmaier** [comment]
I thought about making it a separate box, but I decided against it, because I figured it's probably even more confusing than not, as that box wouldn't actually take folders just "subfolders" into the download directory. Those users who will actually use it will probably either figure it out themselves they can put folders into the mask, or google "how to..." and arrive here (or at any doc we create in the future).

**Comment by kpedro88** [comment]
Fair enough. As a lower-priority item, maybe an option for a separate box (disabled by default) could eventually be added.

**Comment by hwprinz** [comment]
Hello,
Would it be possible to 'read' the browser setting itself, where downloads are to put to....?
so, setting would be either 'user'/download-dir
or
'ask where to put to...'

kind regards

**Comment by nmaier** [comment]
> Would it be possible to 'read' the browser setting itself, where downloads are to put to....?

It is not possible.

**Comment by jdrch** [comment]
> 
> 
> Successfully pulled from multiple websites. was able to use the mask and _batch_ to create directories.
> 
> but for new users the default behavior is that it just dumps to downloads, potentially spamming the users file system.
> 
> could a "download to..." dropdown/preference be added?
> 
> [Bugzilla] https://bugzilla.mozilla.org/show_bug.cgi?id=1342563

Added a [comment](https://bugzilla.mozilla.org/show_bug.cgi?id=1342563#c21) there asking for the feature to be implemented sooner rather than later.

WebExtensions have been an absolute plague. Ugh.

**Comment by Mr-Personality** [comment]
Firstly, I am SO happy that DTA is back after all this time. Thank goodness! I really missed it and had given up hope of it ever returning! :) I've got tons of bookmarked galleries waiting to download, dating back years.

Anyway, I came here to report this bug, that the download path option is missing and DTA saves everything to the desktop.

I find this conversation hard to follow, but from what I can gather, it seems like the author is saying it isn't possible to specify download location? That's confusing as the feature has always been part of DTA?

Without the ability to select download location, the extension would be almost useless. For one thing, it would mean you can only download one gallery at a time. And then after every download, you'd have to go to the desktop and copy the files to where they need to go.

With all due respect, this would be horrible, and nothing like the former greatness of DTA. Why the change? I don't get it.

**Comment by kpedro88** [comment]
Unfortunately, this is a change in Firefox; they have limited the APIs available for WebExtensions. See my comment at https://github.com/downthemall/downthemall/issues/2#issuecomment-526801494 for a user-side workaround (and note that you can put a *relative* path in the "mask" box).

**Comment by nmaier** [comment]
@Mr-Personality DownThemall HAS to use the download folder configured in your browser. You can configure that folder in your browser settings itself (DownThemAll! is not allowed to change it, not even on your behalf, only you can).

However, you can still have it create subfolders, by using the mask.
e.g. you could use `*y*/*m*/*d*/*name*.*ext*` to create folders by date, or `*batch*/*name*.*ext*` to have it create a new numbered subflolder for each time you add new downloads.

Given the user feedback so far, I decided to add back a folder dropdown. However, as previously stated, you cannot select an absolute location with it (as browsers won't let us), but you can specify subfolder within your downloads directory.

**Comment by jdrch** [comment]
Not much you can do about that :(

While I still use DTA on my workstation because I don't want to install an unapproved download manager, I've moved to [Persepolis](https://github.com/persepolisdm/persepolis) (and the associated [Firefox extension](https://addons.mozilla.org/en-US/firefox/addon/persepolis-dlm-integration/)) for non-sniffed downloads on my personal machines.

**Comment by nmaier** [comment]
Ah, aria2.... I contributed some code to aria2 a while back :grin:
So, while this isn't really the place to "advertise" competitors 😝 , that one I don't really mind as it is open source.

**Comment by jdrch** [comment]
> I contributed some code to aria2 a while back

That's pretty cool. It seems pretty powerful, especially if you use it in a script. 

> this isn't really the place to "advertise" competitors

Sorry, that's not what I meant :( Was just trying to provide an alternative for users who want to be able to select arbitrary download folders. As I said, DTA is still useful for content sniffing and on machines on which standalone app installation is restricted.

**Comment by Mr-Personality** [comment]
> However, you can still have DownThemAll create subfolders, by using the mask.
> e.g. you could use `*y*/*m*/*d*/*name*.*ext*` to create folders by date, or `*batch*/*name*.*ext*` to have it create a new numbered subflolder for each time you add new downloads.

Hey, thanks for the reply nmaier :) I didn't realize you could do this. That is really helpful. Not as good as being able to change directory, but still, it means I can now get things done, and use this add-on properly.

I hadn't realized what "Mask" was. To me, the word "mask" does not suggest that you can do this stuff. If it were me, I would make the following changes, to make it easier to use (these are humble suggestions):

1. Split the mask field into two - the top one with the heading "Sub directory" and the bottom one with the heading "File name format". This makes everything clearer to understand, and makes it easier to quickly change the directory or filename string without affecting the other (most of the time you would only want to change one or the other).

2. For the Sub directory field, I would make the default value something like "DTA downloads\" - this would both illustrate how it works, and keep all the downloads in a neat location by default.

3. I didn't know what the tags 'legend' button was at first. I would definitely add a tooltip to this, or better still, change the button to text, so that it reads "Show naming codes…" or "List of naming codes…" (or something like that). I would also move this button directly under (or between) the two fields so it seems more immediately related.

4. The *url* tag is "without protocol". I would make this "without protocol or filename", since including the filename just puts every individual image in its own folder, which is pointless.

I would also like to request a feature…

Instead of the default directory being "DTA downloads\", you make it "DTA downloads\*url portion*\". Then directly below this box, you have the heading "URL portion", followed by the site's full URL excluding protocol and filename. The right most portion of the URL (after the last slash) is in bold, indicating this will be used as the *URL portion*. Then to the right, there are < and > buttons, to change which portion is used. (Hope that made sense?)

**Comment by Mr-Personality** [comment]
> However, you can still have DownThemAll create subfolders, by using the mask.
> e.g. you could use `*y*/*m*/*d*/*name*.*ext*` to create folders by date, or `*batch*/*name*.*ext*` to have it create a new numbered subflolder for each time you add new downloads.

Hey, thanks for the reply nmaier :) I didn't realize you could do this. That is really helpful. Not as good as being able to change directory, but still, it means I can now get things done, and use this add-on properly.

I hadn't realized what "Mask" was. To me, the word "mask" does not suggest that you can do this stuff. If it were me, I would make the following changes, to make it easier to use (these are humble suggestions):

1. Split the mask field into two - the top one with the heading "Sub directory" and the bottom one with the heading "File name format". This makes everything clearer to understand, and makes it easier to quickly change the directory or filename string without affecting the other (most of the time you would only want to change one or the other).

2. For the Sub directory field, I would make the default value something like `"DTA downloads\" `- this would both illustrate how it works, and keep all the downloads in a neat location by default.

3. I didn't know what the tags 'legend' button was at first. I would definitely add a tooltip to this, or better still, change the button to text, so that it reads "Show naming codes…" or "List of naming codes…" (or something like that). I would also move this button directly under (or between) the two fields so it seems more immediately related.

4. The *url* tag is "without protocol". I would make this "without protocol or filename", since including the filename just puts every individual image in its own folder, which is pointless.

I would also like to request a feature…

Instead of the default directory being `"DTA downloads\"`, you make it `"DTA downloads\*url portion*\"`. Then directly below this box, you have the heading "URL portion", followed by the site's full URL excluding protocol and filename. The right most portion of the URL (after the last slash) is in bold, indicating this will be used as the *URL portion*. Then to the right, there are < and > buttons, to change which portion is used. (Hope that made sense?)

**Comment by d7415** [comment]
>     1. Split the mask field into two - the top one with the heading "Sub directory" and the bottom one with the heading "File name format". This makes everything clearer to understand, and makes it easier to quickly change the directory or filename string without affecting the other (most of the time you would only want to change one or the other).

If you look just above your message... https://github.com/downthemall/downthemall/commit/31cb23923ad3400b7b44c2a216e528f40bcb3544

>     2. For the Sub directory field, I would make the default value something like `DTA downloads\ `- this would both illustrate how it works, and keep all the downloads in a neat location by default.

Please no. I think a separate field makes it clear how it works.

>     3. I didn't know what the tags 'legend' button was at first. I would definitely add a tooltip to this, or better still, change the button to text, so that it reads "Show naming codes…" or "List of naming codes…" (or something like that). 

I agree with the tooltip - I knew what it was, but I think it was in the old version of DTA. I'm not convinced about button text as it makes  a small button (which, imo, is best kept small as it shouldn't be needed often) into a big button.

> I would also move this button directly under (or between) the two fields so it seems more immediately related.

Everything on that line (in the latest release anyway) is mask related...

>     4. The `*url*` tag is "without protocol". I would make this "without protocol or filename", since including the filename just puts every individual image in its own folder, which is pointless.

Are you sure about this? I don't have time to test this morning, but that should depend on the rest of the mask - if you just use `*url*` as the mask, they should be stored in the same directory tree as on the server (with an extra layer for the hostname). If you're using `*url*/*name*.*ext*` then yes, that would be pointless, but is not the tag's fault.

> Instead of the default directory being `DTA downloads\`, you make it `DTA downloads\*url portion*\`. Then directly below this box, you have the heading "URL portion", followed by the site's full URL excluding protocol and filename. The right most portion of the URL (after the last slash) is in bold, indicating this will be used as the `*URL portion*`. Then to the right, there are < and > buttons, to change which portion is used. (Hope that made sense?)

This sounds like a variation on #59. The UI sounds clunky though.

**Comment by Mr-Personality** [comment]
> If you look just above your message... [31cb239](https://github.com/downthemall/downthemall/commit/31cb23923ad3400b7b44c2a216e528f40bcb3544)
Thanks, I would not have seen that. I find Git Hub confusing and don't know how to navigate.

> > 4. The `*url*` tag is "without protocol". I would make this "without protocol or filename", since including the filename just puts every individual image in its own folder, which is pointless.
> > ```
> 
> Are you sure about this?
Yeah, it creates a subdirectory for each individual image. No one wants that, surely? 

> > Instead of the default directory being `DTA downloads\`, you make it `DTA downloads\*url portion*\`. Then directly below this box, you have the heading "URL portion", followed by the site's full URL excluding protocol and filename. The right most portion of the URL (after the last slash) is in bold, indicating this will be used as the `*URL portion*`. Then to the right, there are < and > buttons, to change which portion is used. (Hope that made sense?)
> 
> This sounds like a variation on #59. The UI sounds clunky though.
Clunky? It's just the URL (which is helpful to see) with two tiny buttons next to it?

> I should probably add that I'm biased towards a minimal UI, partially because I've been using DTA for years so a) know my way around and b) am slightly resistant to change.

I've been using DTA since the early 2000s, but I care about things being easy to use and easy to understand for the sake of allowing other people to enjoy this great software.

**Comment by d7415** [comment]
> Thanks, I would not have seen that. I find Git Hub really confusing and don't know how to navigate. It's a wonder I was even able to find my way here and post here.

It can be a bit daunting to start with.

> Yeah, it creates a subdirectory for each individual image. No one wants that, surely?

I just tested this and it behaved as expected. If the files are in the same directory on the server, they're saved in the same directory locally.

> Clunky? It's just the URL (which is helpful to see) with two tiny buttons next to it?

Well it would only be the url of one file, so would have to update for that, and may not be the same result for all files. Unless you mean the page url? This interface would also be limited to this tag so wouldn't be very flexible.

While waiting for the new DTA I tried Download Star, which has a "filter" interface that would achieve this (linked in #59). That may not be the best solution or the way @nmaier wants to go, but it *is* flexible. This should really be in a new issue though, as this one is for setting the target folder of downloads.

> I've been using DTA since the early 2000s, but I care about things being easy to use and easy to understand for the sake of allowing other people to enjoy this great software. If the interface is alienating or confusing then fewer newcomers will commit to using it, which ultimately means less interest and support for the software.

True, but I would consider it far more confusing to have 2 buttons and a URL there that aren't necessarily used in the filename or path at all! My experience is that most "normal" users just put everything into their downloads folder, which is probably part of why Mozilla locked extensions out of everywhere else.

**Comment by Mr-Personality** [comment]
> I just tested this and it behaved as expected. If the files are in the same directory on the server, they're saved in the same directory locally.

I've tested it with a few sites and each pic is always saved in its own individual folder (with ".jpg" at the end of the folder name)

> My experience is that most "normal" users just put everything into their downloads folder

I'm not saying you're wrong, but I cannot understand why anyone would do that. That would just create a mess that you'd have to tidy up later.

**Comment by d7415** [comment]
> I've tested it with a few sites and each pic is always saved in its own individual folder (with ".jpg" at the end of the folder name)

What mask were you using? Are you using the new subfolders dropdown? A mask of just `*url*` with no trailing slash should work as I described.

> > Unless you mean the page url?
> 
> Yes, that's what I meant. When I'm saving galleries I generally copy the portion of the gallery's URL that is unique to the gallery. (Normally a bunch of numbers.) I can't think of a better system than this.

Ah. You need to be clearer - `*url*` is the url of the file, not the page. I have no use for that in my masks, but I can see it being useful for others.

> > My experience is that most "normal" users just put everything into their downloads folder
> 
> I'm not saying you're wrong, but I cannot understand why anyone would do that. That would just create a mess that you'd have to tidy up later.

Maybe, but I'd have to tidy up if I was using your scheme, so.... I think a simple default mask is better than a complex one that may lead to "Where are my files? What are all these folders for? I didn't ask for this!".

I think your way should be possible, but I'm not convinced it's a good default. My two cents, anyway.

**Comment by Mr-Personality** [comment]
> What mask were you using? … A mask of just `*url*` with no trailing slash should work as I described.

This mask:
`*url*\*name*.*ext*`

Results in every JPEG having its own individual folder.

> > I cannot understand why anyone would save everything into their downloads folder. That would just create a mess that you'd have to tidy up later.

> Maybe, but I'd have to tidy up if I was using your scheme, so.... 

I don't have to tidy up. Each gallery is saved to where it's supposed to be, and that's where it stays. Or, with this new version of DTA, I have to temporarily save stuff to the desktop then relocate it later, but it's still organized to begin with so there's no tidying up to do.

**Comment by nmaier** [comment]
Or you could `myfoldername/*name*.*ext*` or whatever. Or wait for the next release which will have a subfolder dropdown.

**Comment by d7415** [comment]
> This mask:
> `*url*\*name*.*ext*`
> 
> Results in every JPEG having its own individual folder.

Ok. Yes, that will do that, because you've told DTA to create a folder of the file's URL, which is necessarily different for each file. If you read what I said, and use `*url*` alone as the mask, it will not do that.

> I don't have to tidy up. Each gallery is saved to where it's supposed to be...

Supposed to be by *your* system. Not everyone is you.


Anyway, this has moved from partially related to entirely off-topic for this thread. From your original four items:

1) Done (unreleased)
2) The default target path (in the unreleased version) is apparently `downthemall`, similar to your suggestion
3) Should be a separate issue.
4) Seems to be a misunderstanding. For page URLs, try `*refurl*`. For changing the number of components, refer to #59.

**Comment by nmaier** [comment]
>The default target path (in the unreleased version) is apparently downthemall, similar to your suggestion

Nah, that's just a default suggestion when you open the drop down menu. The default is empty

**Comment by Mr-Personality** [comment]
> > Results in every JPEG having its own individual folder.

> Ok. Yes, that will do that, because you've told DTA to create a folder of the file's URL, which is necessarily different for each file. If you read what I said, and use `*url*` alone as the mask, it will not do that.

I find this totally confusing but after a bit of testing I (think) I get it now. I assumed that `*url*` was exclusively for folder structure and had nothing to do with the filename. But it seems that `*url*` can also affect the filename. It's like the folder structure and filename are being treated as one long string. I have never come across that way of thinking before in any application and it's difficult for me to get my head around it.

**Comment by d7415** [comment]
> It's like the folder structure and filename are being treated as one long string. 

[Yes, that's what a URL is...](https://en.wikipedia.org/wiki/Url)

**Comment by TMPhoenix** [comment]
I would love to have something like this back, even if it is just in a subdirectory of the downloads folder (could use links to break out of the downloads folder anyway). I used to use the context menu list of folders to quickly send downloads to the correct folder. Would love to be able to do that again. It would save me a *LOT* of time.

**Comment by nmaier** [comment]
DownThemAll never had folders in the context menu? Do you mean the dropdown maybe, or are you referring to another extension perhaps?

**Comment by nmaier** [comment]
@TMPhoenix ^ (I always forget to tag people, sorry)

**Comment by MorsInvictaEst** [comment]
I really hope that we'll see the neccessary API changes being implemented rather sooner than later, assuming they will happen at all. In the meantime, I've tried using symlinks to my most important download destinations under the default download directory. I'm happy to say that thanks to the implementation of the subfolder drop-down, I now have at least part of the old functionality back.
Maybe this might be useful as a stop-gap solution for advanced users while we wait for better solutions.

**Comment by BugMagnet** [comment]
I am a little lost on this.
My firefox option is to ask me every time what download folder to use.
Why isn't dta allowed to use my preference setting and ask me where to put the files?

**Comment by TMPhoenix** [comment]
@nmaier, I'm fairly certain the old version had a way to send files to a particular location from a list of recently used folders without going through the whole save as... > open file save dialog > browse for folder > click ok route... Maybe I'm misremembering it being in the context menu, but I know it did not require opening any additional file save dialogs. 

Hmmm, maybe it was from the save as dialog, where you choose to open/save, but I imagine that extensions may have a hard time extending that now with additional options.

**Comment by nmaier** [comment]
> Why isn't dta allowed to use my preference setting and ask me where to put the files?

@BugMagnet Only the creators of the WebExtensions system know (which technically is Google Chrome folks, since mozilla only copied them). The idea is probably that extensions shouldn't be able to place files in arbitrary locations in the system or even overwrite arbitrary files. Which makes kinda sense. But then again, keeping a per-extension list of allowed folders that users can specify wouldn't have been too hard for browsers to implement (but of course, browsers didn't implement anything like that). So extensions are stuck with only being allowed to write files to the Downloads folder.

>Hmmm, maybe it was from the save as dialog, where you choose to open/save, but I imagine that extensions may have a hard time extending that now with additional options.

@TMPhoenix Not "hard time". Outright impossible.

**Comment by icq4ever** [comment]
I tried with 'ln -s' in linux. and it works :)

**Comment by planestranger** [comment]
I was so happy to find out DTA was resurrected. Yet inability to use absolute paths completely breaks my workflow that includes downloading files to different disks during one session. It's not DTA I was keeping my outdated Firefox version for. Is there any possibility to make a plugin for some other modern, regularly updated browser like... Vivaldi for example... or something ... that won't have such limitations?

**Comment by nmaier** [comment]
@planestranger 
>Yet inability to use absolute paths completely breaks my workflow that includes downloading files to different disks during one session.

Yes, I hate it too. Since worst WebExtensions limitation in my book and the browsers really fucked this one up. Well, Chrome fucked it up and then Firefox blindly copied it.

>Is there any possibility to make a plugin for some other modern, regularly updated browser like... Vivaldi for example... or something ... that won't have such limitations?

**Vivaldi, Brave, Opera, etc**  are all Chromium shells. While they could in theory offer  an extended/modified extension system, they don't. So the same limitations that apply to Chrome/Chromium apply to them too. **Firefox**, I already mentioned, blindly copied the WebExtensions API from Chrome, so for now it has the same limitations in the downloads API. **Safari** has their own system, but again based on Chrome and Apple cripples it more and more instead of making it better, recently. **MS Edge** is now based on Chromium too, and like Vivaldi etc inherited the same limitations from it. The old **non-Chromium MS Edge** had the WebExtensions API as well, or rather parts of it, it didn't allow custom downloads at all.

So that's about it for modern browsers that allow extensions. And none is an option for this particular issue.

Workarounds
===
Basically, it comes down to "linking" other folders into the downloads folder.

*nix, incl Linux, BSDs, macOS:
---
You can create symlinks (or mount points). `ln -s <from> $HOME/Downloads/<to>` in short.

Windows
---
Windows actually supports symbolic links, and junction points (soft links) since at least Windows 7, but it hides it well, on NTFS file systems at least. Creating symbolic links seem to require elevated privileges, while junction points do not (junction points can only point to directories, which is fine, and only on the local computer, which may or may not be fine)
Essentially, `mklink /J drive_d D:\` would create such a junction point to D:\. Symlinks can be made with`mklink /D drive_d D:\`.
There are guides and third parties tool on how to create and manage these things on the web.

**Comment by samxyzabc** [comment]
Can you just open a save location selection dialog, as before, and make a relative path to the download location settled in the browser settings for users? I thought this is a workaround and very handful and is technically possible.

**Comment by nmaier** [comment]
@samxyzabc we'd have to open one such dialog for every single download, which isn't a sane way to deal with this mess either

**Comment by fencepost-error** [comment]
There's another workaround possible for Windows, on the basis of, as a coworker of mine used to put it, "if they can't get their s**t together maybe we can get it together for them".  If GoogleZilla don't want to allow you to save things anywhere other than the Downloads folder then the solution is to move the Downloads folder to where you need it to be.  Go to My Computer, right-click Downloads, Properties, Location, and then select "Move...".  You can now relocate your downloads folder to wherever you want it to be.  Test it by using DTA to grab something off a web page, it'll now appear where you want it to, not where GoogleZilla want it to.

As a use case for any GoogleZilla folks reading this (based on the bugzilla thread where the reaction seems to be "Why would anyone want to save anything anywhere other than the system Downloads folder?"), we have students download containers and VM images for assignments, and they can't go into the local downloads folder because they're too big, they have to be saved to a (huge) network share that gets scrubbed after a few days.  Since Firefox seems to be incapable of downloading anything more than a hundred MB or two without the download breaking or failing in some way, and since DTA and similar managers are crippled by the WebExtensions API, we came up with this workaround.  Unfortunately it still causes problems because students need to download things like assignments into a permanent download location, not a temporary storage area on a network share.  In other words they need at least two distinct locations for downloads, and GoogleZilla has chosen to make it impossible to do this.

**Comment by smnthermes** [comment]
@nmaier but you were asked to maintain the extension for Pale Moon and then declined. Stockholm syndrome much?

DownThemAll 3.0.8 works properly on PM. However, many people spread FUD about PM (debunked by [1](https://old.reddit.com/r/palemoon/comments/evywwj/how_secure_is_pm_really/ffzc5iw/), [2](https://www.palemoon.org/releasenotes.shtml), [3](https://www.palemoon.org/roadmap.shtml), [4](https://forum.palemoon.org/viewtopic.php?f=65&t=22399) and [5](https://forum.palemoon.org/viewtopic.php?f=65&t=22270)), which halts the development of add-ons. I hate Google!

**Comment by shadree** [comment]
I'm not very code savvy so I apologise if this has been answered already here but I've been using DTA with Firefox for years now, and there is an option on the download prompt "Save files in:" where I can add a download location per download. Then when I use OneClick, it will send all subsequent downloads to that location.
Now I'm trying out Brave and I don't know how to add the option.
I think it may have been part of FlashGot which I can't see on the Chrome Web Store.
![DTA-FF-prompt](https://user-images.githubusercontent.com/33830456/75667419-8ce0a000-5cc7-11ea-92fe-06f297741597.png)

**Comment by fencepost-error** [comment]
That's Firefox Classic with XUL, before Mozilla broke it.  Try any recent version and you'll find it's been disabled.

**Comment by shadree** [comment]
My follow-up question is: can it be implemented in the current version(s)?

**Comment by d7415** [comment]
> That's Firefox Classic with XUL, before Mozilla broke it. Try any recent version and you'll find it's been disabled.

There's a "Subfolder" option in recent versions

> My follow-up question is: can it be implemented in the current version(s)?

The current version should have a "Subfolder" option. You can only choose locations within your default download directory (that's a limitation of webextensions) but, as discussed [earlier](https://github.com/downthemall/downthemall/issues/2#issuecomment-557751776) in this issue there are workarounds.

**Comment by LucaNonato** [comment]
I confirm that the symlink "trick" also works on macOS, but there's an issue: when i open the "DTA! Manager" the saved files are shown as "Missing" even if they are actually present in the folder in which DTA saved them. Am i the only one? Using Firefox 76 with macoS 10.13.6

**Comment by nmaier** [comment]
> when i open the "DTA! Manager" the saved files are shown as "Missing" even if they are actually present in the folder in which DTA saved them. Am i the only one? Using Firefox 76 with macoS 10.13.6

This is probably a side effect of the browser simply "forgetting" the downloads. Firefox will put old downloads into an archived state, at which point DownThemAll! cannot get any information about them from the browser again. This is a limitation in Firefox/WebExtensions and nothing I can fix.

**Comment by nixsee** [comment]
Sorry, but I can't figure out how to use the masks for either subfolders or filename mask. I have tried *text*.*ext* and it just keeps it as whatever the filename is, and despite using various combinations of *domain*, *subdirs* and *host* with /  and \ as separators, all files just go in the main download directory. I'm using it on Microsoft Edge.

**Comment by BEEDELLROKEJULIANLOCKHART** [comment]
I have configured Firefox to ask me where to download files to. Is lack of adherence to this preference one thing that DownThemAll is not able to control? If so, should I report this to Firefox as erroneousness?

**Comment by nl255** [comment]
Unfortunately it is not possible in Firefox to configure a download directory if you have it set to ask you where you want to save files to.  So you need to set it to always save files to a specific folder, download whatever you want with downthemall, then change the setting back to ask when you are done.  Yes, it is a royal pain in the ass and you are probably better off using something like Xtreme Download Manager if possible.

**Comment by UTer** [comment]
In Firefox, this application ignores the "Always ask you where to save files" setting parameter to select the directory.

**Comment by ACGLINT** [comment]
heyo duno if this was written anywhere else but for anyone still looking for a work around to this is to use folder symbolic links, its a built in windows feature you can do through the cmd line but someone also made a thing called Link Shell Extension that adds it to the right click menu as well, https://schinagl.priv.at/nt/hardlinkshellext/linkshellextension.html

Please read both steps before clicking and deleting folders, you only want to delete the subfolder thats made by the extension first.

1# what i did first was create a subfolder using the download manager pop up and then putting in the name of folder i want it to go to, this can also even be on a different drive. for example lets say you have a folder called music123 on the D drive (It wasnt clear to me at first but the option is at the bottom with the drop down menu and the default one is downthemall.) type that name in and press download and you will find that new subfolder and it will now be an option in that list, you can delete this new music123 folder afterwards because you only need the option for it in the list.

2# the next step is to create a symbolic link using the LSE (ngl i cant be asked to explain it the cmd line way its long and also someone else could explain it here instead or look it up your self.), lets go back to that music123 folder you have on the D drive, right click the folder and pick the fancy new "pick link source" option, then go back to the location/directory where you just deleted the empty music123 folder right click an empty part of the file list and a new fancy "drop as" list should appear, click the symbolic link option.

if all things were done and explained right you should have a fancy new folder called music123 with a lil green shortcut icon instead of the blue one, next time you download a song or whatever and choose the music123 in the download manager list it will then go through the shortcut and save to the main folder wherever it may be. wonder if it works with cloud folders as well.

sorry if i didnt explain it right, but if i did go have fun downloading stuff

**Comment by woodenphone** [comment]
I workaround this restriction by using symlinks / Junction Points from `"${Env:USERPROFILE}/Downloads"` / `"%USERPROFILE%/Downloads"`   to the dir(s) that I want to actually write to.

e.g. Downloads dir is on my OS SSD where space is expensive and want to save to a HDD where space is cheap.

---

All of the following is copypasted from my personal notes:

Log:
```cmd
C:\Users\User>mklink /d C:\Users\User\Downloads\dta  E:\dta
symbolic link created for C:\Users\User\Downloads\dta <<===>> E:\dta
```

## Links
* [mklink command  (learn.microsoft.com))](https://learn.microsoft.com/en-us/windows-server/administration/windows-commands/mklink)
* [ command  (learn.microsoft.com))]()
* [Symlinks in Windows 10! (Windows Developer Blog)](https://blogs.windows.com/windowsdeveloper/2016/12/02/symlinks-windows-10/)
* [New-Item - "Example 7: Create a symbolic link to a file or folder" (Powershell docs, learn.microsoft.com)](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.management/new-item?view=powershell-7.3#example-7-create-a-symbolic-link-to-a-file-or-folder)
* ["The Complete Guide to Creating Symbolic Links (aka Symlinks) on Windows" (howtogeek.com)](https://www.howtogeek.com/16226/complete-guide-to-symbolic-links-symlinks-on-windows-or-linux/)
* [Where is the "downloads" folder located (stackoverflow.com)](https://stackoverflow.com/questions/57947150/where-is-the-downloads-folder-located)
* ["Hard links and junctions" (learn.microsoft.com)](https://learn.microsoft.com/en-us/windows/win32/fileio/hard-links-and-junctions)
